CID-18090: update contact form

taimoor-zafar-nhs opened the pull request:
**Description:**
This PR corresponds to [CID-18090 ](https://nhsd-jira.digital.nhs.uk/browse/CID-18090)JIRA ticket. It updates the help centre [Contact Us form](https://help.dev.signin.nhs.uk/contact). 

**For Reviewers:**
All Help Centre environments point to Live ZenDesk, so submitting a contact form will result in a real ticket being raised.  To avoid sending test entries into Live it's possible to test the contact form by entering a specific line of text in the Message field.  In ZenDesk this will be recognised as coming from a test and will then auto close the entry and respond to the users email. The message to enter in the Describe the problem field should be: 

"Contact Us Form submitted for NHS login regression test"

- [X] Screen scores 100% on lighthouse report
- [X] Trigger Zendesk query
- [X] Dynamic character count of 1500 characters 


## Previous Form:
<img width="1721" alt="previous-contact-form" src="https://github.com/NHSDigital/nhs-login-help-centre/assets/130666551/d8261b52-405d-41f7-a2e1-1052801df1c6">

## Updated Form:
<img width="1072" alt="new-contact-form" src="https://github.com/NHSDigital/nhs-login-help-centre/assets/130666551/1675eb43-9651-469b-9dc0-e9ba47867bf4">

ramuned-NHSD commented on src/contact.njk:
```diff
@@ -20,48 +19,28 @@ excludeFromSearch: true
-
+                <div class="nhsuk-inset-text">
+                    <h2 class="nhsuk-heading-m">We cannot help with medical problems</h2>
+                    <p>This form is only for help using NHS login</p>
```
Missing full stop

ramuned-NHSD commented on src/contact.njk:
```diff
@@ -20,48 +19,28 @@ excludeFromSearch: true
+                    </p>
+                    <p class="nhsuk-body">Call 999 if it's a life-threatening emergency.</p>
+                </div>
+                {# Name input field #}
```
Why do we have this and email one on line 41? The id's and text already indicate what the fields are for, do we need these?

taimoor-zafar-nhs replied:
Those were only for additional clarity to easily find the elements during development. They are removed now.

ramuned-NHSD commented on src/contact.njk:
```diff
@@ -20,48 +19,28 @@ excludeFromSearch: true
+                    <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                         Email address
                     </label>
                     <span class="nhsuk-hint">
```
Can't comment but the line below this - `This should be the email address you use for NHS login.` no longer has a full stop in the prototype

jacobcallear commented:
I don't think the form validation error messages are working properly: I only get the proper red error messages if I've selected options for "Select the website or app you are trying to visit" and "Select the problem you are having". Also the error message for an empty message in "Describe the problem in more detail" doesn't show the error in the box at the top any more? I don't know if the design has changed though

ramuned-NHSD commented on src/js/contact-page.js:
```diff
@@ -4,10 +4,77 @@
+    clientName = client.value;
+  });
+
+  dupeAccountRadioButton.addEventListener('change', function(){
+    selectedMessage = SELECTED_MESSAGES_MAPPING[dupeAccountRadioButton.value];
+  });
+
+  gpConnectionIssueRadioButton.addEventListener('change', function(){
+    selectedMessage = SELECTED_MESSAGES_MAPPING[gpConnectionIssueRadioButton.value];
+  });
+
+  changeEmailRadioButton.addEventListener('change', function(){
+    selectedMessage = SELECTED_MESSAGES_MAPPING[changeEmailRadioButton.value];
+  });
+
+  notSureRadioButton.addEventListener('change', function(){
+    selectedMessage = SELECTED_MESSAGES_MAPPING[notSureRadioButton.value];
+  });
```
I'm probably missing something, why do we need to be setting this selectedMessage? Can't spot it being used anywhere

taimoor-zafar-nhs replied:
Good catch! This is the common issue i.e new radio buttons introduced above the message field. It was needed to be concatenated with the text from the 'message' field. I've added it to the message text now.
Changing the variable name to `selectedCommonIssue` instead of `selectedMessage`.

ramuned-NHSD replied:
Aah didn't realise, there wasn't a mention of needing to add radio button description to the free text when sending in the Jira ticket.

ramuned-NHSD commented on src/js/contact-page.js:
```diff
@@ -4,10 +4,77 @@
+  var clientName = "";
+  var selectedMessage = "";
+
+  visitNHSAppRadioButton.addEventListener('change', function(){
+    clientListDropdown.classList.add("nhsuk-radios__conditional--hidden");
+    clientName = visitNHSAppRadioButton.value;
+  });
+
+  visitPatientAccessRadioButton.addEventListener('change', function(){
+    clientListDropdown.classList.add("nhsuk-radios__conditional--hidden");
+    clientName = visitPatientAccessRadioButton.value;
+  });
+
+  visitNHSPrescriptionRadioButton.addEventListener('change', function(){
+    clientListDropdown.classList.add("nhsuk-radios__conditional--hidden");
+    clientName = visitNHSPrescriptionRadioButton.value;
+  });
+
+  visitOtherRadioButton.addEventListener('change', function(){
+    clientListDropdown.classList.remove("nhsuk-radios__conditional--hidden");
+    client.setAttribute("required", "")
+  });
+
+  client.addEventListener('change', function(){
+    clientName = client.value;
+  });
```
Curious, why do we need all these event listeners? Is the formData not picking up the selected value otherwise?

taimoor-zafar-nhs replied:
The client list dropdown needs to rendered conditionally into DOM i.e when "Other" radio button is selected.
The change event listener on the radio buttons doesn't catch the deselection event that is why I've added these event listernes to show/hide the client list dropdown and to catch the appropriate client name. 

I can only think of 1 other way to do this (add a single event listener to the radio group and handle these events conditionally) but that will complicate the logic by adding conditions. I am happy to do it differently just cannot think of any better way.

jacobcallear commented on src/contact.njk:
```diff
@@ -20,48 +19,28 @@ excludeFromSearch: true
                         Full name
                     </label>
                     <span class="nhsuk-error-message nhs-help-centre__form-control-error"></span>
                     <input class="nhsuk-input nhsuk-u-width-one-half nhs-help-centre__form-control-input" id="name"
```
The prototype has the same width for this input box and the one for email address

taimoor-zafar-nhs commented:
> I don't think the form validation error messages are working properly: I only get the proper red error messages if I've selected options for "Select the website or app you are trying to visit" and "Select the problem you are having". Also the error message for an empty message in "Describe the problem in more detail" doesn't show the error in the box at the top any more? I don't know if the design has changed though

@jacobcallear I've addressed the empty message issue. 🚀 
The challenge with displaying red error messages for the 'Select the website or app' and 'Select the problem' radio groups lies in the inability of the radio group to detect whether any of its radio items have been selected. I know this can be done in Angular but here I am not able to it. As a workaround, I've added a 'required' validation which ensures that users must select at least one radio items.

taimoor-zafar-nhs commented on src/js/contact-page.js:
```diff
@@ -3,11 +3,60 @@
+  const commonProblems = document.getElementById('common-problems');
+  const clientLisElement = document.getElementById('clients-list');
+  const clientDropdown = document.getElementById('client');
+  var clientName = "";
```
It is declared `var` because it's value is updated in a scope of a different function i.e event listeners.
Declaring it as `let` would restrict it to the same scope.
